python/lsst/meas/algorithms/defects.py

```diff
@@ -151,6 +151,25 @@ def __eq__(self, other):
     def __str__(self):
         return "Defects(" + ",".join(str(d.getBBox()) for d in self) + ")"
 
+    def normalize(self, region):
+        """Recalculate defects bounding boxes to use minimal set.
+
+        Parameters
+        ----------
+        region : `lsst.geom.Box2I
+            Region of the exposure for which the defects are found.
+
+        Returns
+        -------
+        normalizedDefects : `lsst.meas.algorithms.Defect`
+            Normalized list of defects.
+        """
+        mi = lsst.afw.image.MaskedImageF(region)
+        self.maskPixels(mi, maskName="BAD")
+        normalizedDefects = Defects.fromMask(mi, "BAD")
+
+        return normalizedDefects
```

Comment:
I wonder, given that this is returning an object, whether it should be changed to be a class-method? Either than, or make it normalize the defects themselves.
But I think I'd like to defer to someone who knows more about both this object and good programming in general, so pinging @timj for his opinion on this.

Reply:
It can't be a classmethod because you are using `self`.

Reply:
@mfisherlevine and I were wondering if it should or should not be a class method? Does it matter in this case?

Reply:
I don't understand. If it's a class method then you would have to pass in the defects you want to normalize as a parameter:

```
normalized = Defects.normalize(myDefects)
```
It seems more natural to me for you to be normalizing the defects you already have:
```
normalized = myDefects.normalize()
```

I'm okay with it return a new `Defects` rather than modifying the existing one. We should consider whether the defects should always be normalized when being saved using writeText or writeFits. That seems like a question that can only be answered from the science case -- maybe you do sometimes want a way to see how you constructed the defects.

Reply:
It seems surprising to me that this returns a new object rather than modifying in place; compare (this implementation of) `Defects.normalize()` with `list.sort()`.

If this is really the behaviour we want (why?), then consider:

- Making this clear in the docstring. This is not just “recalculating defect bounding boxes”, it is “returning a normalized copy” (see the docstring for `Defects.transpose()`).
- Personally, I'd call the method “Defects.normalized()” for clarity, but that's not a hard requirement (and may not be consistent with the existing `transpose()` API).

By my reading of the code (not having actually tried it), `defects.normalize() != defects` in general (the bounding boxes *may* have changed). If that's the case, I don't think you can automatically normalize on write, regardless of the science case: if would be super weird if `defects.readFits` gave me back something different from the data I'd just written.

Reply:
I was torn about the mutability. I had assumed initially that `normalize` would change in place (so after the fromMask call you'd replace the bounding boxes in `self` with the new bounding boxes).  People tend to prefer new objects be returned though. You make a good case that `transpose` returns a new object so there is precedent for that. The counter argument is that before and after calling normalize the effective defects are the same.  You are right that `transpose` not being `transposed` makes prior art for this not being called `normalized`. As I said, I'm torn between new object and mutability (it's not like `Defects` are immutable though given that you can add new bounding boxes at any time).

Regarding defects != defects.normalize -- at some level they probably should compare equal. It depends on the semantics of what we mean by equality. In most cases "convert to the same pixel mask" is what we mean by equality and that's not what happens at the moment. We could make that be true by comparing the normalized versions.

I'm fine with write writing whatever it has without normalizing.

Reply:
I'd be quite tempted to maintain it as an invariant that the defects are normalized rather than exposing it to the user as part of the API. Perhaps with a special context for bulk-updates. Something like (excuse my toy API):

```python
class Defects(object):
    def __init__(self):
        # ...
        self._bulk_mode = False

    def _normalize(self):
        # ...

    def append(self, defect):
        self._defectList.append(defect)
        if not self._bulk_mode:
            self._normalize()

    @contextmanager
    def bulk_mode(self):
        self._bulk_mode = True
        try:
            yield
        finally:
            self._bulk_mode = False
            self._normalize()
```

Perhaps that would be too disruptive a change at this point, though.

Reply:
That's a really interesting idea. Having Defects always be normalized behind the scenes would make equality much more useful and the context manager stops it being slow in a loop (and we'd have to use the context manager when reading from file).